### PR TITLE
Added MoveUp/MoveDown commands

### PR DESCRIPTION
**Pull request by TIHan**
This adds move up and move down gestures for F# projects.

Original issue here: https://github.com/dotnet/project-system/issues/2887

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Microsoft.VisualStudio.ProjectSystem.FSharp.VS.csproj [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project Sdk="Microsoft.NET.Sdk">
-  <Import Project="..\VisualStudio.props"/>
+  <Import Project="..\VisualStudio.props" />
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Microsoft.VisualStudio.ProjectSystem.FSharp.VS.csproj]: Nit: no meaningful changes in this file - revert the changes?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/xlf/Menus.vsct.cs.xlf [hunk 2]
```diff
@@ -7,44 +7,64 @@
...
       </trans-unit>
+      <trans-unit id="DebugTargetMenuDebugFrameworkMenu|ButtonText">
+        <source>Framework</source>
+        <target state="translated">Architektura</target>
+        <note />
+      </trans-unit>
       <trans-unit id="cmdidGenerateNuGetPackageProjectContextMenu|ButtonText">
         <source>&amp;Pack</source>
-        <target state="translated">&amp;Zabalit</target>
+        <target state="new">&amp;Pack</target>
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/xlf/Menus.vsct.cs.xlf]: Something is not right here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+using System.Threading.Tasks;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Packaging;
+using Microsoft.VisualStudio.ProjectSystem.Input;
+using Microsoft.VisualStudio.Shell;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    [ProjectCommand(ManagedProjectSystemPackage.ManagedProjectSystemOrderCommandSet, ManagedProjectSystemPackage.MoveDownCmdId)]
+    [AppliesTo(ProjectCapability.FSharp)]
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs]: Could these (and the vsct file, etc), just be in the FSharp dll?  Any reason for having them here?

File: src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Microsoft.VisualStudio.ProjectSystem.FSharp.VS.csproj (hunk 1 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Microsoft.VisualStudio.ProjectSystem.FSharp.VS.csproj]: Yea. Looks like it auto-formatted it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/xlf/Menus.vsct.cs.xlf (hunk 2 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/xlf/Menus.vsct.cs.xlf]: Indeed. This is wrong.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs (hunk 3 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs]: I haven't tried moving those into the Managed.FSharp.VS project. But, in Managed.VS we have the CSharp and VB specific commands together, ex: AddClassProjectCSharpCommand.cs, AddClassProjectFSharpCommand.cs.

From that I assumed it was ok to have the commands in Managed.VS.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.TestServices/ProjectSystem/ProjectTreeParser/ProjectTreeParser.cs [hunk 4]
```diff
@@ -266,12 +266,27 @@ private void ReadFields(MutableProjectTree tree)
...
                         throw _tokenizer.FormatException(ProjectTreeFormatError.UnrecognizedPropertyName, $"Expected 'FilePath', 'Icon' or 'ExpandedIcon', but encountered '{fieldName}'.");
                 }
             }
         }
 
+        private void ReadDisplayOrder(MutableProjectTree tree)
+        {
+            Tokenizer tokenizer = Tokenizer(Delimiters.PropertyValue);
+
+            var ident = tokenizer.ReadIdentifier(IdentifierParseOptions.None);
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.TestServices/ProjectSystem/ProjectTreeParser/ProjectTreeParser.cs]: >ident [](start = 16, length = 5)

can we just call this `identifier`? We tend to not abbreviate in this repo

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs (hunk 3 above)
Comment (Pilchie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs]: It's not the *worst*, it just means that dll is bigger, and takes up address space for all projects, even though they will never be applicable, etc.

It also can make it harder to find things.

I haven't been around project-system long enough to know if there is some reason this needs to be like this though.  Maybe someone else will tell me this is right.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.TestServices/ProjectSystem/ProjectTreeParser/ProjectTreeParser.cs (hunk 4 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.TestServices/ProjectSystem/ProjectTreeParser/ProjectTreeParser.cs]: Sure.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs (hunk 3 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs]: I agree it should be in FSharp.VS, perhaps I can try to see what happens :)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractMoveCommand.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+            IVsUIShell shell = serviceProvider.GetService(typeof(SVsUIShell)) as IVsUIShell;
+
+            if (shell == null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            object pvar = null;
+            IVsWindowFrame frame = null;
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractMoveCommand.cs]: fame declaration can be inlined `shell.FindToolWindow(0, ref persistenceSlot, out var frame)`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    internal static class OrderingHelper
+    {
+        private enum MoveAction { Up=0, Down }
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Can we explicitly set `Down=1`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        /// <summary>
+        /// Gets the display order for a project tree.
+        /// </summary>
+        private static int GetDisplayOrder(IProjectTree projectTree)
+        {
+            if (projectTree is IProjectTree2 projectTree2)
+            {
+                return projectTree2.DisplayOrder;
+            }
+            return 0;
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Can you explain why it is safe to return `0` here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractMoveCommand.cs (hunk 5 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractMoveCommand.cs]: Yup.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 6 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Sure.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+                    parent.InsertBeforeChild(element, referenceElement);
+                    break;
+
+                case MoveAction.Down:
+                    parent.InsertAfterChild(element, referenceElement);
+                    break;
+            }
+        }
+
+        private static async Task<bool> Move(ConfiguredProject configuredProject, IProjectTree projectTree, MoveAction moveAction)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Consider naming this `MoveAsync`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 7 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I'll add a comment. It's safe to return because the display order value of nodes that are not set explicitly are always zero.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+        {
+            var projectLockService = configuredProject.UnconfiguredProject.ProjectService.Services.ProjectLockService;
+
+            // Do a write lock.
+            using (var access = await projectLockService.WriteLockAsync())
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: name this something more descriptive like `writeLock`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 8 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: That's fair. In that case I'll rename the others to Async as well.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 9 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: That's fine. Do note that almost call use cases of using `WriteLockAsync` call it `access`.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+        {
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return GetSiblingByMoveAction(projectTree, MoveAction.Up) != null;
+        }
+
+        /// <summary>
+        /// Move the project tree up over one of it's sibling.
+        /// </summary>
+        public static Task<bool> MoveUp(ConfiguredProject configuredProject, IProjectTree projectTree)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Consider renaming to `MoveUpAsync`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        {
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return GetSiblingByMoveAction(projectTree, MoveAction.Down) != null;
+        }
+
+        /// <summary>
+        /// Move the project tree down over one of it's sibling.
+        /// </summary>
+        public static Task<bool> MoveDown(ConfiguredProject configuredProject, IProjectTree projectTree)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Consider renaming to `MoveDownAsync`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 10 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Yes

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 11 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Yes

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs (hunk 3 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/MoveDownCommand.cs]: Yea, maybe we should discuss this part more.

**Comment by TIHan** [comment]
This is almost done. I added more tests with moving up and down files and then checking the xml results.

The last part is dealing with the xlf translations which I think are broken :(.

**Comment by Pilchie** [comment]
@tmeschter to help with the xlf stuff.  I suspect you could revert it all, and run another build?  I wonder if something got out of sync with a merge in the middle of when it ran.

**Comment by tmeschter** [comment]
@TIHan Try this:

1. Revert the changes to the .xlf files (and only the .xlf files).
2. Rebuild (this will cause the .xlf files to be updated).
3. Commit the updated .xlf files.
4. Push the changes so I can see where that leaves us. :-)

If we end up with the same weird result, there is likely a bug in the XLIFF MSBuild task.

**Comment by TIHan** [comment]
@tmeschter ok, that seemed to have worked! :)

This is declared finished now.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+            Requires.NotNull(project, nameof(project));
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return Move(unconfiguredProject, project, projectTree, MoveAction.Down);
+        }
+
+        /// <summary>
+        /// Move the project tree down over one of its siblings.
+        /// </summary>
+        public static Task<bool> MoveDownAsync(ConfiguredProject configuredProject, IProjectTree projectTree)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Can we re-order these to be public -> private so the entry point into this class is at the top?

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Yup.

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    internal static class OrderingHelper
+    {
+        private enum MoveAction { Up=0, Down=1 }
+
+        private static void ProcessProjectTree(Queue<IProjectTree> treeQueue, HashSet<string> lookup, UnconfiguredProject unconfiguredProject, IProjectTree projectTree)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I don't like that this method modifies both `treeQueue` and `lookup` but returns `void`.  Perhaps return a tuple so its clear what the inputs and outputs are?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+        {
+            Requires.NotNull(configuredProject, nameof(configuredProject));
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return MoveAsync(configuredProject, projectTree, MoveAction.Down);
+        }
+
+        private enum MoveAction { Up=0, Down=1 }
+
+        private static void ProcessProjectTree(Queue<IProjectTree> treeQueue, HashSet<string> lookup, UnconfiguredProject unconfiguredProject, IProjectTree projectTree)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I don't like that this method modifies both `treeQueue` and `lookup` but returns `void`. Perhaps return a tuple so its clear what the inputs and outputs are?

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I'll see what I can do.

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Done.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+        {
+            Requires.NotNull(configuredProject, nameof(configuredProject));
+            Requires.NotNull(projectTree, nameof(projectTree));
+
+            return MoveAsync(configuredProject, projectTree, MoveAction.Down);
+        }
+
+        private enum MoveAction { Up=0, Down=1 }
+
+        private static HashSet<string> CreateRootedPathLookup(UnconfiguredProject unconfiguredProject, Project project, IProjectTree projectTree)
```
Comment (jmarolf) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: add `summary` comments?

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Done.

**Comment by TIHan** [comment]
Ok, I did another round of commenting and changing a few method names.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+        private static IProjectTree2 GetPreviousSibling(IProjectTree projectTree)
+        {
+            return GetSiblingByDisplayOrder(projectTree, (i, orderedChildren) =>
+            {
+                if (i == 0)
+                {
+                    return null;
+                }
+
+                return orderedChildren.ElementAt(i - 1) as IProjectTree2;
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Does this `ElementAt` end up using an indexer, or does it cause *another* walk through the tree?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+                // The reference element is the element we want to insert above or below.
+                var referenceElement = GetReferenceElementByMoveAction(unconfiguredProject, project, sibling, moveAction);
+
+                if (referenceElement != null && referenceElement.Parent != null)
+                {
+                    // If the project tree we are wanting to move is a folder, we might do many insertions and removals.
+                    if (projectTree.IsFolder)
+                    {
+                        var elements = GetElements(unconfiguredProject, project, projectTree);
+                        if (elements.Count() == 0)
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Prefer `!elements.Any()` rather than forcing a linear walk of the whole collection if it isn't something that is type tested for a Count/Length.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+                }
+            }
+        }
+
+        /// <summary>
+        /// Moves child elements based on the reference element and move action.
+        /// </summary>
+        private static void MoveChildrenByMoveAction(IEnumerable<ProjectItemElement> elements, ProjectItemElement referenceElement, MoveAction moveAction)
+        {
+            // If we are moving elements down, we need to reverse order.
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: *why*?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+
+            if (sibling != null)
+            {
+                // The reference element is the element we want to insert above or below.
+                var referenceElement = GetReferenceElementByMoveAction(unconfiguredProject, project, sibling, moveAction);
+
+                if (referenceElement != null && referenceElement.Parent != null)
+                {
+                    // If the project tree we are wanting to move is a folder, we might do many insertions and removals.
+                    if (projectTree.IsFolder)
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Consider eliminating this check, and just always doing the multiple case, which will end up being the same in the single item case anyway.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+
+                case MoveAction.Down:
+                    return GetNextSibling(projectTree);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Gets a reference element based on the given project tree and move action. Can return null.
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I have no idea what a "reference element" means.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 21]
```diff
@@ -0,0 +1,361 @@
...
+        /// Get project item elements based on the project tree.
+        /// Project tree can be a folder or item.
+        /// </summary>
+        private static IEnumerable<ProjectItemElement> GetElements(UnconfiguredProject unconfiguredProject, Project project, IProjectTree projectTree)
+        {
+            var lookup = CreateRootedPathLookup(unconfiguredProject, project, projectTree);
+
+            var elements = new List<ProjectItemElement>();
+
+            foreach (var item in project.AllEvaluatedItems)
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Isn't this a pretty huge collection?  Can it be filtered to only `Compile` items or anything?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 16 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: It will use an index. https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/ElementAt.cs#L28

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 20 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I described what it was in TryMove. I'll move the description here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 21 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: It potentially can be, but I can't just make Compile items be the only thing we care about because we also order non-compilable items too.

In CPS though, they do filter by having an instance of `IProjectVersionedValue<IProjectItemSchema>` and getting a list of item types that we do care about: `var itemTypes = schemaData.Value.GetKnownItemTypes();` then they `project.AllEvaluatedItems.Where(item => itemTypes.Contains(item.ItemType))`.

Let me see what can be done here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 18 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: If I don't reverse the elements before calling `InsertAfterChild`, the end result will reverse the ordering of the elements that were part of the folder.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 19 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: That's fair. Will make this even simpler.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 21 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I managed to find a much better way to handle this. The new approach simplified these methods as well.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+                    foreach (var childTree in tree.Children)
+                    {
+                        treeQueue.Enqueue(childTree);
+                    }
+                }
+                else
+                {
+                    if (tree is IProjectItemTree2 tree2 && IsValidDisplayOrder(tree2.DisplayOrder))
+                    {
+                        includes.Add(tree2.Item.ItemName);
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: This is unique enough?

**Comment by Pilchie** [comment]
Needs to be rebased onto `dev15.6.x`.

Then, if you are confident in the uniqueness, it can be merged.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs (hunk 22 above)
Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: It's the evaluated include, which I assume so. Though, thinking about it, it might be possible there could be two, but it's unlikely.

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: I need to do a little bit more work on this bit. Includes need to be in order.

Comment (TIHan) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/OrderingHelper.cs]: Ok, I have a sanity check for uniqueness, and the includes are now in order by their display order.

**Comment by TIHan** [comment]
Ok, I rebased, but I assume we want to merge it into dev15.6.x branch then?

**Comment by jmarolf** [comment]
🎊
